site/src/lib/model/apiKey.model.ts

```diff
@@ -0,0 +1,239 @@
+  ApiKeyProperties,
+  "hashedString" | "salt" | "user"
+>;
+
+type ApiKeyDocument = ApiKeyProperties;
+
+export class ApiKey {
+  createdAt: Date;
+  displayName: string;
+  hashedString: string;
+  lastUsedAt?: Date | null;
+  publicId: string;
+  revokedAt?: Date | null;
+  salt: string;
+  useCount: number;
+  user: DBRef;
+
+  static readonly COLLECTION_NAME = "bp-api-keys";
+
+  private static readonly HASHING_SECRET = process.env.HASHING_SECRET;
+
+  private static readonly PREFIX = "b10ck5";
+
+  private static readonly KEY_FORMAT_REGEXP = new RegExp(
+    // e.g. b10ck5.e84a616cedc8822b02ac96761b068f19.91ea0c90-acee-4215-a01e-85a4438af0ea
+    `^${ApiKey.PREFIX}\\.[a-z0-9]{32}\\.[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$`,
+  );
+
+  private constructor({
+    createdAt,
+    displayName,
+    hashedString,
+    lastUsedAt,
+    publicId,
+    revokedAt,
+    salt,
+    useCount,
+    user,
+  }: ApiKeyProperties) {
+    this.createdAt = createdAt;
+    this.displayName = displayName;
+    this.hashedString = hashedString;
+    this.lastUsedAt = lastUsedAt;
+    this.publicId = publicId;
+    this.revokedAt = revokedAt;
+    this.salt = salt;
+    this.useCount = useCount;
+    this.user = user;
+  }
+
+  private static hash({
+    publicId,
+    privateId,
+    salt,
+  }: {
+    publicId: string;
+    privateId: string;
+    salt: string;
+  }) {
+    if (!this.HASHING_SECRET) {
+      throw new Error(
+        "Cannot hash API key: no HASHING_SECRET present in environment.",
+      );
+    }
+
+    const stringToHash = `${publicId}.${privateId}.${salt}`;
+
+    return crypto
+      .createHmac("sha256", this.HASHING_SECRET)
+      .update(stringToHash)
+      .digest("hex");
+  }
+
+  static async create(
+    db: Db,
+    params: { displayName: string; user: User },
+  ): Promise<string> {
+    if (!this.HASHING_SECRET) {
+      throw new Error(
+        "Cannot create API key: no HASHING_SECRET present in environment.",
+      );
+    }
+
+    const { displayName, user } = params;
+
+    const privateId = uuid();
+    const publicId = crypto.randomBytes(16).toString("hex");
+    const salt = crypto.randomBytes(32).toString("hex");
+
+    const hashedString = this.hash({ privateId, publicId, salt });
+
+    const apiKey = new ApiKey({
+      createdAt: new Date(),
+      displayName,
+      hashedString,
+      publicId,
+      salt,
+      useCount: 0,
+      user: user.toRef(),
+    });
+
+    await db.collection<ApiKeyDocument>(this.COLLECTION_NAME).insertOne(apiKey);
+
+    return `${this.PREFIX}.${publicId}.${privateId}`;
+  }
+
+  private static async getByPublicId(
+    db: Db,
+    params: { publicId: string },
+  ): Promise<ApiKey | undefined> {
+    const apiKeyRecord = await db
+      .collection<ApiKeyDocument>(this.COLLECTION_NAME)
+      .findOne({
+        publicId: params.publicId,
+      });
+
+    if (!apiKeyRecord) {
+      return undefined;
+    }
+
+    return new ApiKey(apiKeyRecord);
```

Comment:
Just for consistency, I've been returning `null` instead when something isn't found in the other model classes:
```suggestion
    return apiKeyRecord ? new ApiKey(apiKeyRecord) : null;
```